wandb/io_wrap.py

```diff
@@ -73,6 +79,81 @@ def write(self, data):
         self.destination.write(data)
 
 
+class WindowSizeChangeHandler(object):
+    """A SIGWINCH handler that keeps a list of FDs to update any time the
+    window size changes.
+
+    This is a singleton initialized by init_sigwinch_handler().
+    """
+    def __init__(self):
+        self.fds = []
+        # bind one of these so we can compare instances to each other
+        self._handler = self._handle_function
+
+    def register(self):
+        try:
+            old_handler = signal.signal(signal.SIGWINCH, self._handler)
+        except ValueError:  # windows
+            logger.warn('Setting SIGWINCH handler failed')
+        else:
+            if old_handler is not None:
+                logger.warn('SIGWINCH handler was not None: %r', old_handler)
+
+    def unregister(self):
+        try:
+            old_handler = signal.signal(signal.SIGWINCH, None)
+        except ValueError:  # windows
+            logger.warn('Setting SIGWINCH handler failed')
+        else:
+            if old_handler is not self._handler:
+                logger.warn('SIGWINCH handler was not from W&B: %r', old_handler)
+
+    def add_fd(self, fd):
+        self.fds.append(fd)
+        self._set_win_sizes()
+
+    def _handle_function(self, signum, frame):
+        try:
+            self._set_win_sizes()
+        except:
+            logger.exception('Exception during SIGWINCH')
+
+    def _set_win_sizes(self):
+        win_size = fcntl.ioctl(sys.stdin.fileno(), termios.TIOCGWINSZ, '\0' * 8)
+        rows, cols, xpix, ypix = array.array('h', win_size)
+        if cols == 0:
+            cols = 80
+        win_size = struct.pack("HHHH", rows, cols, xpix, ypix)
+        for fd in self.fds:
+            fcntl.ioctl(fd, termios.TIOCSWINSZ, win_size)
+
+
+def init_sigwinch_handler():
+    global SIGWINCH_HANDLER
+    if SIGWINCH_HANDLER is None and sys.platform != "win32" and not os.environ.get('WANDB_TEST') and not os.environ.get('WANDB_DEBUG'):
+        SIGWINCH_HANDLER = WindowSizeChangeHandler()
+        SIGWINCH_HANDLER.register()
+
+
+def wandb_pty(resize=True):
+    """Get a PTY set to raw mode and registered to hear about window size changes.
+    """
+    master_fd, slave_fd = pty.openpty()
+    # raw mode so carriage returns etc. don't get added by the terminal driver,
+    # bash for windows blows up on this so we catch the error and do nothing
+    # TODO(adrian): (when) will this be called on windows?
+    try:
+        tty.setraw(master_fd)
```

Comment:
This seems pretty different, will it change output for users?

Reply:
This has always been there (although I refactored the code). By default, TTYs "cook" the data that passes through them, handling CRs, LFs, etc. -- it's actually exactly what we'd want to do for the backend or whatever. Anyway, we want to disable the "cooking" in this case (I think) because *our* CR/LF-handling code assumed uncooked input.